### DISCUSSION TITLE
Page Hooks vs Action Hooks

### DISCUSSION THREAD

**Discussion by ecrmnn** [filamentphp/filament#14738] [Help]
### Package

Panel builder

### Package Version

v3.2

### How can we help you?

Hi! 

I'm a seasoned Laravel developer, but fairly new to Filament.

I have a few questions about hooks.
I might be completely off here, so go easy on me.

### In my project I have two resources with Pages:
- Department
- Employee

In addition to this, Department has an EmployeeRelationManager.
I render the related employees on the edit page for a given Department.

When I create, update or delete an employee, I need to make an API-call to a third party service.

I want CRUD operations to work the same on Pages and RelationManager (modal).

I would expect that the default `CreateAction` would simply run the logic implemented in my Page, so that I didnt have to implement it twice. Or at least that they used the same hooks.
But I've had to implement the logic twice, since the Pages and Actions hooks are two different beasts.

### Example
In `CreateEmployee.php` I can use this hook
```php
class CreateEmployee extends CreateRecord
{
    protected function handleRecordCreation(array $data): Model
    {
       // Custom logic
    }
}
```

But this does not affect the modal, so I'll have to implement it there too:
In `EmployeeRelationManager.php` I'll have to implement the same logic, but now, in the `action` hook:
```php
class EmployeeRelationManager extends RelationManager
{
    public function table(Table $table): Table
    {
        return $table
            ->recordTitleAttribute('name')
            ->columns([
                Tables\Columns\TextColumn::make('name'),
            ])
            ->filters([
                //
            ])
            ->headerActions([
                Tables\Actions\CreateAction::make()
                    ->action(function (array $data) {
                        // Custom logic
                    }),
            ])
            ->actions([
                //
            ])
            ->bulkActions([
                //
            ]);
    }
}
```

I've ran into other scenarios where I'd want to use the `beforeValidate` hook in my Page, then I have to reimplement this in my Action, but there I need to use the `beforeFormValidated` hook.

This quicky becomes very complex given that the hooks have different names, arguments and even return types.

I might be completely off track here.
Am I doing something really wrong, or is this the way?

Appreciate any feedback on this 🙏

**Comment by leandrocfe** [comment]
Maybe you could create a `Pivot Model` and use [Laravel Observers](https://laravel.com/docs/11.x/eloquent#defining-observers) to handle this?

**Reply by ecrmnn** [reply]
Definitely! That's a great idea.
I guess I got stuck by trying to find _The Filament Way_ of doing it that I missed the obvious.

Still find it a bit weird that the naming of hooks is so inconsistent.
But maybe it's not how you're suppose to solve this kind of things.
